engine/src/main/java/org/hibernate/search/engine/mapper/schema/management/SchemaElementContext.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.engine.mapper.schema.management;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Optional;
+
+import org.hibernate.search.util.common.annotation.Incubating;
+
+@Incubating
+public interface SchemaElementContext {
+	enum OutputFormat {
+		/**
+		 * Generate a JSON representation of this context.
+		 */
+		SIMPLE;
+	}
+
+	/**
+	 * @return The name of the backend to which the index this context represents belongs to.
+	 * If it is a part of the default backend - {@link Optional#empty() an empty optional} will be returned.
+	 */
+	Optional<String> backendName();
+
+	/**
+	 * @return The name of the index this context represents.
+	 */
+	String indexName();
```

Comment:
The downside of this compared to my visitor is that it doesn't work well with exports that are not strictly related to the *index* schema. Imagine we export the scripts to initialize a Kafka stream for Debezium, for example; what would `indexName` return?

That's why I think instead of a `Consumer<SchemaElementContext>`, you should expose some kind of visitor interface such as the `SearchSchemaCollector`/`IndexSchemaCollector` in my initial proposal.

The good news is, it's totally compatible with your concept of "unwrappable"/"extendable" `SchemaElementContext`.

Reply:
Right... I think that I've missed the point that this collector would be used to walk through different "schemas". That's why it took me a different path 😃 
Now, if I got you correctly - we want to have this general-purpose schema walker that we'll extend in the future by adding more visit methods. This would apply to `SearchSchemaManager`/`PojoScopeSchemaManager` while the currently underlying `IndexSchemaManager`s are fine with just providing the "element" that we'll feed into this collector since the index schema manager is .. well for index schema 😃 

If that's what we are looking for - I'll clean things up before you do the proper review 🙈😃

Reply:
Yes that's the idea :100: